keps/262-ConfigurableFailurePolicy/README.md

```diff
@@ -0,0 +1,377 @@
+spec:
+  # Failure Policy configured to restart any failed child job in the cpu-worker ReplicatedJob, but fall back to the
+  # default behavior of restarting the entire JobSet if any child job in the tpu-worker ReplicatedJob fails.
+  failurePolicy:
+    rules:
+    - action: RestartJob
+      targetReplicatedJobs: cpu-worker
+    maxRestarts: 10
+  replicatedJobs:
+  - name: tpu-worker 
+    replicas: 4 # set to number of node pools
+    template:
+      spec:
+        parallelism: 2 # number of nodes per pool
+        completions: 2 # number of nodes per pool
+        backoffLimit: 0
+        template:
+          spec:
+            nodeSelector:
+              cloud.google.com/gke-tpu-accelerator: tpu-v4-podslice
+              cloud.google.com/gke-tpu-topology: 2x2x2
+            containers:
+            - name: main
+              image: pytorch:latest
+              command: ["..."]
+  - name: cpu-worker 
+    replicas: 1 # set to number of node pools
+    template:
+      spec:
+        parallelism: 1 # number of nodes per pool
+        completions: 1 # number of nodes per pool
+        backoffLimit: 0
+        template:
+          spec:
+            restartPolicy: Never
+            containers:
+            - name: main
+              image: python:3.10
+              command: ["..."]
+              
+```
+
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Proposed Failure Policy API:
+
+```go
+
+// FailurePolicyAction defines the action the JobSet controller will take for
+// a given FailurePolicyRule.
+type FailurePolicyAction string
+
+const (
+	// Fail JobSet immediately, regardless of maxRestarts.
+	FailJobSet FailurePolicyAction = "FailJobSet"
+
+	// Don't count the failure against maxRestarts.
+	Ignore FailurePolicyAction = "Ignore"
+
+	// The failed child job is permanently failed and should not be
+	// restarted, active child jobs will continue to run to completion.
+	FailJob FailurePolicyAction = "FailJob"
+
+	// Restart the failed child job only, not the whole JobSet.
+	RestartJob FailurePolicyAction = "RestartJob"
+)
+
+// FailurePolicyRule defines a FailurePolicyAction to be executed if a child job
+// fails due to a reason listed in OnJobFailureReasons.
+// Only only rule can match a unique (Job failure reason, ReplicatedJob) pair.
+type FailurePolicyRule struct {
```

Comment:
Check indentation on this API. Its a bit uneven.

Reply:
Fixed.